addons/account/models/product.py

```diff
@@ -118,6 +118,28 @@ def _check_uom_not_in_invoice(self):
                 "If you want to change its Unit of Measure, please archive this product and create a new one."
             ))
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        products = super().create(vals_list)
+        if products.company_id:
+            return products
+        # If no company was set for the product, the product will be available for all companies and therefore should
+        # have the default taxes of the other companies as well. sudo() is used since we're going to need to fetch all
+        # the other companies default taxes which the user may not have access to.
+        other_companies = self.env['res.company'].sudo().search([('id', '!=', self.env.company.id)])
+        if not other_companies:
+            return products
+
+        default_customer_tax_ids = other_companies.filtered('account_sale_tax_id').account_sale_tax_id.ids
+        default_supplier_tax_ids = other_companies.filtered('account_purchase_tax_id').account_purchase_tax_id.ids
+
+        for product in products.sudo():
+            if self._context.get('default_taxes_id', True):
+                product.taxes_id = [Command.link(tax) for tax in default_customer_tax_ids]
+            if self._context.get('default_supplier_taxes_id', True):
+                product.supplier_taxes_id = [Command.link(tax) for tax in default_supplier_tax_ids]
```

Comment:
Correct me if I'm wrong but doesn't the context key `default_[field_name]=False` mean that we shouldn't get the default values for that field while creating a record? So the logic here actually just makes sure that the key value is not False if it exists, to make sure we can add the default taxes of the rest of the companies.



> self.env.context

What's the difference?

Reply:
Well, basically, both the `default_` key in the context set to a falsy value, and any write value given are expected to cancel the default behavior.
In this situation, I don't really know what could be the best approach honestly.  We could consider that we don't need to add taxes for any of the environment companies, and change the default to set all the taxes of the current companies :thinking:.

It's tricky to make sure that we do not add unexpected taxes that the customer wouldn't want.

Reply:
> We could consider that we don't need to add taxes for any of the environment companies, and change the default to set all the taxes of the current companies

Not sure I understand, can you elaborate more please?

Reply:
Currently, on the field, the default is `lambda self: self.env.company.account_sale_tax_id or self.env.company.root_id.account_sale_tax_id`.

We could say that if you have multiple companies enabled in the company switcher, we add the taxes for all the companies by default: `lambda self: self.env.companies.account_sale_tax_id or self.env.companies.root_id.account_sale_tax_id`, and in your create override, you would only add the defaults for companies not chosen in the switcher
`other_companies = self.env['res.company'].sudo().search([('id', 'not in', self.env.companies.ids)])`

Reply:
I had already discussed this with @chklop and it was decided not to change the default on the field as it would over-complicate things and also having multiple companies enabled in the company switcher is not going to be super common :sweat_smile:

Reply:
I think multi-company is intrinsically a complex thing and half magic solutions may lead to more problems than they solve, but well, let's start with the simple solution then.